File: web/lib/ticker.js
// Emit events on a named class's occurrences (twin of ticker.py).
//
// Epoch-anchored and drift-free: every timeout is recomputed from the
// current clock against the shared timeline, so two browsers running the
// same name fire in lockstep (± NTP error ± timer jitter).
//
//   import { ticker } from "./lib/ticker.js";
//   const t = ticker(phase(Frac.parse("1/3"), 3, {k: 2}),
//                    (pulse) => console.log("state 2 began", pulse.start));
//   t.stop();

import { nextPulse, phase, PhaseClass, PSet, Windowed } from "./core.js";
import { nowTai } from "./timescale.js";
import { Frac, F } from "./rat.js";
import { parse } from "./names.js";

// EventEmitter style: a browser/Node-native EventTarget that dispatches a
// "pulse" CustomEvent on every occurrence.
//
//   const em = new PhaseEmitter(cls);
//   em.addEventListener("pulse", (e) => console.log(e.detail.start));
//   em.stop();
export class PhaseEmitter extends EventTarget {
  constructor(cls, opts = {}) {
    super();
    this.cls = cls;
    this._t = ticker(cls, (pulse) => {
      this.dispatchEvent(new CustomEvent("pulse", { detail: pulse }));
    }, opts);
  }
  stop() {
    this._t.stop();
  }
}

// Async-iterator style: `for await (const pulse of pulses(cls)) ...`
export async function* pulses(cls, { lead = 0 } = {}) {
  let lastFired = null;
  for (;;) {
    const t = nowTai();
    const q = lastFired && lastFired.cmp(t) > 0 ? lastFired : t;
    const nxt = nextOccurrence(cls, q); // strictly after q — no duplicates
    if (!nxt) return;
    const delayMs = nxt.start.sub(t).toNumber() * 1000 - lead;
    if (delayMs > 0) await new Promise((r) => setTimeout(r, delayMs));
    lastFired = nxt.start;
    yield nxt;
  }
}

// Support-aware next: a Windowed class stops emitting past its support.
function nextOccurrence(cls, t) {
  if (cls instanceof Windowed) {
    let p = nextPulse(cls.cls, t);
    while (p && p.end.cmp(cls.support.start) <= 0) p = nextPulse(cls.cls, p.start);
    if (!p || p.start.cmp(cls.support.end) >= 0) return null;
    return p;
  }
  return nextPulse(cls, t);
}

// setInterval, but the interval is a NAME on the shared timeline.
//
//   const h = setIntervalClock(() => console.log("tick"), "1/3");     // every 1/3 s
//   const h2 = setIntervalClock(fn, 250);                              // every 250 ms
//   const h3 = setIntervalClock(fn, "ic1:c:w=1/3;m=3;phi=2/3");        // a named class
//   clearIntervalClock(h);
//
// Unlike setInterval, firings land on the epoch-anchored lattice: aligned
// across every machine running the same spec, and drift-free (each wait is
// re-derived from the clock). A number is milliseconds for familiarity; a
// string is either a rational period in seconds ("1/3") or an ic1: name.
export function setIntervalClock(callback, spec, opts = {}) {
  let cls;
  if (typeof spec === "number") {
    const P = new Frac(BigInt(Math.round(spec)), 1000n);
    cls = phase(P.div(F(2n)), 2, {});
  } else if (typeof spec === "string") {
    if (spec.startsWith("ic1:")) {
      cls = parse(spec);
    } else {
      const P = Frac.parse(spec);
      cls = phase(P.div(F(2n)), 2, {});
    }
  } else {
    cls = spec;
  }
  if (!(cls instanceof PhaseClass || cls instanceof PSet || cls instanceof Windowed)) {
    throw new Error("setIntervalClock needs a periodic class, a period, or ms");
  }
  return ticker(cls, callback, opts);
}

export function clearIntervalClock(handle) {
  handle?.stop?.();
}

export function ticker(cls, callback, { lead = 0 } = {}) {
  let stopped = false;
  let timer = null;
  let lastFired = null; // guard: timers can fire early; never re-fire a pulse

  function arm() {
    if (stopped) return;
    const t = nowTai();
    const q = lastFired && lastFired.cmp(t) > 0 ? lastFired : t;
    const nxt = nextOccurrence(cls, q); // strictly after q — no duplicates
    if (!nxt) return; // no further occurrences
    const delayMs = nxt.start.sub(t).toNumber() * 1000 - lead;
    timer = setTimeout(() => {
      if (stopped) return;
      lastFired = nxt.start;
      callback(nxt);
      arm(); // re-derive from the clock — never accumulate drift
    }, Math.max(delayMs, 0));
  }

  arm();
  return {
    stop() {
      stopped = true;
      clearTimeout(timer);
    },
  };
}

File: python/src/intervalclock/ticker.py
"""Emit events on a named class's occurrences.

Epoch-anchored: each wait is computed fresh from the shared timeline
(next_after of the *current* clock reading), so there is no accumulated
drift, and two NTP-disciplined machines running the same name fire in
lockstep to within their sync error. Works for phase classes, PSets,
windowed classes, and cron schedules alike — anything with a next
occurrence.

    import asyncio, intervalclock as ic

    async def main():
        async for span in ic.ticks(ic.phase(ic.rat("1/3"), 3, k=2)):
            print("pulse", span)          # fires every 1 s, at state 2

    asyncio.run(main())

or with a callback:

    stop = ic.every("*/5 * * * *", lambda span: print("cron fired", span))
"""

from __future__ import annotations

import asyncio
import threading
from typing import AsyncIterator, Callable, Optional

from .core import Span, next_after
from .cron import CronSchedule, cron_next_after
from .timescale import Instant, now


def _next(x, t):
    if isinstance(x, CronSchedule):
        return cron_next_after(x, t)
    return next_after(x, t)


async def ticks(x, *, lead: float = 0.0) -> AsyncIterator[Span]:
    """Async-iterate the occurrences of x as they happen.

    Yields each occurrence Span at (its start − lead) seconds. Ends when the
    class has no further occurrences (e.g. a windowed class past its
    support).
    """
    last = None  # timers can wake early; never yield the same span twice
    while True:
        t, _ = now()
        q = t if last is None or last <= t.t else Instant(last)
        nxt = _next(x, q)
        if nxt is None:
            return
        delay = float(nxt.start - t.t) - lead
        if delay > 0:
            await asyncio.sleep(delay)
        last = nxt.start
        yield nxt


def iter_ticks(x, *, lead: float = 0.0):
    """Plain blocking generator — `for span in iter_ticks(cls): ...`.

    The synchronous twin of ticks() for scripts and REPLs: sleeps until each
    occurrence and yields it.
    """
    import time as _time

    last = None
    while True:
        t, _ = now()
        q = t if last is None or last <= t.t else Instant(last)
        nxt = _next(x, q)
        if nxt is None:
            return
        delay = float(nxt.start - t.t) - lead
        if delay > 0:
            _time.sleep(delay)
        last = nxt.start
        yield nxt


def every(x, callback: Callable[[Span], None], *, daemon: bool = True) -> Callable[[], None]:
    """Fire callback(span) on each occurrence of x from a background thread.

    Returns a zero-argument stop function. The wait is re-derived from the
    clock before every firing (epoch-anchored, drift-free).
    """
    stop_event = threading.Event()

    def loop():
        last = None
        while not stop_event.is_set():
            t, _ = now()
            q = t if last is None or last <= t.t else Instant(last)
            nxt = _next(x, q)
            if nxt is None:
                return
            delay = float(nxt.start - t.t)
            if stop_event.wait(timeout=max(delay, 0.0)):
                return
            last = nxt.start
            callback(nxt)

    thread = threading.Thread(target=loop, daemon=daemon, name="intervalclock-ticker")
    thread.start()

    def stop():
        stop_event.set()
        thread.join(timeout=1)

    return stop
